### DISCUSSION TITLE
Manual type hints resolves to partially unknown since some releases

### DISCUSSION THREAD

**Discussion by alkatar21** [microsoft/pyright#5150] [Q&A] [answered]
I'm not quite sure, but as far as I understand the template this is the right place and not the bug report?

I use pyright through pylance. I have always explicitly assigned a type to variables whose type could not be infered. But since a short time (probably some weeks) this seems to suppress the error and the type is not only taken as the assigned type, but also as "Unknown". Is this a bug or a wanted new behavior? If it is intentional, how to tell pyright/pylance now that this is the type of the variable?
```py
#!/usr/bin/env python3
# pyright: strict

import tensorflow as tf  # pyright: ignore[reportMissingTypeStubs]

# This worked before.
expected: tf.Tensor = tf.constant(result) # Error: Type of "expected" is partially unknown  Type of "expected" is "Unknown | Tensor"

device: tf.config.PhysicalDevice  # This worked before to typehint device in the following line.
for device in tf.config.list_physical_devices('GPU'): # Error: Type of "device" is partially unknown. Type of "device" is "Unknown | PhysicalDevice"
    tf.config.experimental.set_memory_growth(device, True)
```

**Comment by erictraut** [comment, marked as answer]
This is intentional behavior, not a bug.

You are telling pyright to use "strict" type checking mode. In this mode, any use of expressions that have `Unknown` types results in an error. 

I recently fixed a bug that caused some uses of `Unknown` values to go unreported.

The code sample you've pasted above is incomplete (i.e. it references symbols that are not defined, such as `result`), so I can't say for sure what's causing the issue in this case, but it's likely due to incomplete type information in the tensorflow library. If you're using libraries whose type information is incomplete, strict mode isn't going to work well for you. I recommend sticking with "basic" type checking mode in this case.

> how to tell pyright/pylance now that this is the type of the variable?

You are already telling pyright which type your variable _must_ contain. You're then assigning a value to that variable, and pyright is unable to tell for sure whether it is type compatible with your variable's declared type. For example, if you declare that a variable is of type `list[str] | list[int]`, and then you assign `[1, 2]` to that variable, pyright can verify that this value (whose type is `list[int]`) does not violate the declared type of your variable. But if you assign a value of type `list[Unknown]`, that may or may not be compatible with `list[str] | list[int]`. When you enable strict type checking mode, pyright will tell you about this because it could represent a hole in static type checking.

**Reply by alkatar21** [reply]
`result` I have actually forgotten to exchange. It is something like `[[1, 2]]`.

But this is only one example for some libraries which unfortunately does not provide type information. I than annotated the variables myself and until recently the variable was also considered as this type by `# pyright: strict` and `Unknown` did not show up. (So I annotated the `Unknown` types myself and with that the `Unknown` was replaced by the type and so I could resolve everything almost normally with strict checking).
This worked very well when it was where or before the variable was first set (like the examples above) and I had assumed that this was intentional and wanted behavior to help pyright when typhints were incorrect or missing. But then I guess that was a bug that has now been fixed?

And as I understand the answer correctly there was never actually this possibility and it was not meant that way and I have to find another solution for it if I want to continue using strict?

**Reply by erictraut** [reply]
When you annotate the type for a variable, you are declaring what type it is allowed to contain. For example, if you specify `y: T`, you're indicating that the variable `y` can be assigned values of type `T` or subtypes thereof. If you then assign a value to `y`, it's the job of a type checker to verify that the type of the assigned value is compatible with `T`. For example, `y: T = my_function()`. If `my_function` returns a type that is not compatible with `T`, a type checker will report the type violation. But what if `my_function` is unannotated and its return type cannot be determined? Then its return type will be `Unknown`. A type checker can either assume that it's OK (which it does in "basic" mode) or potentially not OK (which it does in "strict" mode). If you're using a library that is not annotated, you're likely to see some `Unknown` types, and "strict" mode won't work well. Your options in this case are:
1. See if you can upgrade the library to a newer version that includes type information.
2. Find a type stub library that supplies the types for the library.
3. Create your own type stub library that includes types for the classes and methods that your program uses from the library. (This can be a lot of work.)
4. Disable "strict" mode and stick with "basic" type checking. You can enable a subset of the strict-mode checks if you would like. Refer to [this table](https://microsoft.github.io/pyright/#/configuration?id=diagnostic-rule-defaults) for a list of all diagnostic rules and their default severity level for "basic" and "strict" mode. The rules that start with "reportUnknown..." are the ones that you will probably want to leave disabled if you don't have complete type information.
5. Use a `cast` call to manually specify the type of an expression that you're assigning to a variable. For example, `y = cast(T, my_function()`. I recommend that you use casts very sparingly because they make your code hard to read and maintain, and they override any type checking that a static type checker would normally perform.

**Reply by alkatar21** [reply]
Thanks for the detailed explanation. The setup is actually a bit more complicated and uses several of the options already.
The exact behavior of `Unknown` in strict mode was actually not so clear to me before, since it has now worked for a long time that way, but now it is of course more correct. Then I will probably gradually change the code bases, so that the messages go down again.
